### ISSUE TITLE
Metrics server

### ISSUE THREAD

**Issue by strowk** [kelseyhightower/kubernetes-the-hard-way#505]
I think it makes sense to add a few chapters for configuring
https://github.com/kubernetes-sigs/metrics-server/
as without it `kubectl top pods` shows errors.

Issues, preventing from using metrics-server as it explained in doc above: 
1) apiserver cannot reach pod network
it's easily solvable by installing kube-proxy on controllers, we can just modify how control plane is bootstrapped, see https://github.com/kelseyhightower/kubernetes-the-hard-way/issues/486 for details

2) apiserver is not configured to authenticate to extensions
Necessary configurations are explained here 
https://kubernetes.io/docs/tasks/access-kubernetes-api/configure-aggregation-layer/
Basically I just needed to add options to kube-apiserver run:
```
  --proxy-client-key-file=/var/lib/kubernetes/admin-key.pem \         
  --proxy-client-cert-file=/var/lib/kubernetes/admin.pem \                                                                    
  --requestheader-client-ca-file=/var/lib/kubernetes/ca.pem \
  --requestheader-allowed-names=admin \  
```
and scp admin-key.pem and admin.pem to controllers. 

I'm not sure though if it's a good idea to use admin user for that, or better to create additional just for this purpose.

**Comment by bryfry** [comment]
> admin-key.pem and admin.pem to controllers.

This may be a bad choice. It will work because admin cert will have the RBAC to work as the proxy client to the metrics server, but it also has many more privileges.  I'm working through getting an appropriately scoped certificate and will report back on what I find!

**Comment by bryfry** [comment]
The [kubernetes certificates best practices documentation](https://kubernetes.io/docs/setup/best-practices/certificates/) says the CN for the proxy should be `front-proxy-client`

Default CN | recommended key path | recommended cert path | command | key argument | cert argument
-- | -- | -- | -- | -- | --
front-proxy-client | front-proxy-client.key | front-proxy-client.crt | kube-apiserver | –proxy-client-key-file | –proxy-client-cert-file

**Comment by bryfry** [comment]
I can confirm that generating a front-proxy-cert and key with the appropriate CA and then passing the following headers works for least-privilege access for the kube-apiserver acting as a proxy to the metrics-server .  Hope this helps!

```
--requestheader-client-ca-file=<CA> \
--requestheader-allowed-names="" \
--requestheader-extra-headers-prefix=X-Remote-Extra- \
--requestheader-group-headers=X-Remote-Group \
--requestheader-username-headers=X-Remote-User \
--proxy-client-cert-file=<front-proxy-client.crt> \
--proxy-client-key-file=<front-proxy-client.key> \
```

**Comment by gawbul** [comment]
What `ca.conf` did you use for the `kube-front-proxy` certs @bryfry?

**Comment by gawbul** [comment]
> What `ca.conf` did you use for the `kube-front-proxy` certs [@bryfry](https://github.com/bryfry)?

Ah, I got it working with the following:

```ini
[front-proxy-client]
distinguished_name = front-proxy-client_distinguished_name
prompt             = no
req_extensions     = front-proxy-client_req_extensions

[front-proxy-client_req_extensions]
basicConstraints     = CA:FALSE
extendedKeyUsage     = clientAuth, serverAuth
keyUsage             = critical, digitalSignature, keyEncipherment
nsCertType           = client
nsComment            = "Front Proxy Client Certificate"
subjectAltName       = @front-proxy-client_alt_names
subjectKeyIdentifier = hash

[front-proxy-client_alt_names]
IP.0  = 127.0.0.1
IP.1  = 192.168.50.4
DNS.0 = front-proxy-client

[front-proxy-client_distinguished_name]
CN = front-proxy-client
C  = UK
ST = North Yorkshire
L  = York
```
